stdin commands for The Hosts

TTtie opened the pull request:
Implements the stdin commands into GAB.

TTtie committed:
The start

TTtie committed:
[eslint|CLI] Add the base listener + dispose function

TTtie committed:
[CLI] Add the stringified constant; we're not done yet!

TTtie commented:
ignore this pr for now; we still working on it

vladfrangu commented on Modules/CLI.js:
```diff
@@ -0,0 +1,35 @@
+class CLI {
+	constructor (db, client) {
+		this.db = db;
```
Database is in the globals.

TTtie commented:
will fix

KingDGrizzle commented:
Multiple complaints:
1. Client when this will run on the sharder?
2. That is not how you do jsdocs comments.

TTtie commented:
:thinking:

TTtie commented:
I may have solutions on these:
1. We may require() this off the sharder or just on the first shard
2. My bad for `{CLI]`

KingDGrizzle commented:
1. I'd say allow us to choose the shard and then, and then run commands on it.
2. No. You don't get it. Those are not jsdoc comments, those start with `/**`

GilbertGobbels commented:
I disagree with point 1.

STDIN is only captured on the master process as the streams of the workers are only accessible by IPC.
I recommend we use IPC in command functions when needed to contact shards, apart from that, this module should only run on the master process.

TTtie committed:
[CLI] Let's work on it

TTtie committed:
[CLI] Let's work on it

TTtie committed:
[Internals/Errors] Resolve some conflicts

TTtie committed:
Merge branch 'indev-4.0.2' into feature/indev-4.0.2/cli-commands

TTtie commented:
github can you show me some wuv

KingDGrizzle commented:
Please use rebase when merging branches.

TTtie committed:
merge with the current gab

TTtie committed:
no u

TTtie committed:
[eslint]

GilbertGobbels commented:
I'd like to repeat that the STDIN is not reachable from any shards. This means that you can only listen for incoming commands in the master process, so it's currently pointless to require a Discord Client object in the module.

TTtie commented:
:ok_hand: Got it  
Will tamper with that later.

TTtie commented:
@GilbertGobbels Should we at least, require the sharder object or no-no?

TTtie committed:
let's get some Premium Support:tm:

TTtie committed:
let's get some Premium Support:tm:

vitalkanev commented on Configurations/auth.js:
```diff
@@ -1,8 +1,8 @@
-		clientToken: "",
+		clientID: "381171142230147075",
+		clientSecret: "SUktHmXqgz_6-cYYSnc98kNMNQKTljI7",
+		clientToken: "MzgxMTcxMTQyMjMwMTQ3MDc1.DRLYlg.AvvNziQHWev06pZ10K6yKv0MqBw",
```
**PLEASE REMOVE TOKEN**

TTtie replied:
OH FUCK  
That's testing bot anyway, so

TTtie commented:
note: the stdin thing is actually listening to your keyboard inputs.  
I may make it listening to the input after all shards get ready

TTtie commented:
let's force push it later, now will work on resetting my fuck ups

TTtie commented:
Reverted the commits.

TTtie committed:
We got the Premium Support:tm:

TTtie committed:
Merge branch 'indev-4.0.2' of https://github.com/gilbertgobbels/gawesomebot into feature/indev-4.0.2/cli-commands

TTtie committed:
[Bugfix|Discord.js] fix someone's fuck up

TTtie committed:
[Commands/CLI] Few additions and a help command

Commit:
String updates

Commit:
More string updates

(For the sake of consistency, only use the verb "to use" when instructing a user to run a command!)

Commit:
Even more string updates

Commit:
[Encryption] Fix some small  bugs

Commit:
[Events/Message] String Update + Massive bug fix
Since we support aliases, command usage would save the usage depending on what command you entered
So, for instance: `ping` may have an alias of `pong`. It would save `ping` if you run `ping` and `pong` if you ran `pong`
This fixes that!

Commit:
Typo

Commit:
[Commands] Make reload a shared command!
Also some web. Sorry Gilberto

Commit:
Wrong Check for reload

Commit:
Nleh, mistype

Commit:
Even more "even more string updates"

Commit:
Some more string updates

Commit:
[IPC] Fix slight JSON error

Commit:
[Bugs] Fix random not being a function

Commit:
[Events/MessageCreate] Fix undefined Constants, alias bugs, and random not being a function

TTtie committed:
[CLI] Add a note to multiline commands + switch to CTRL+G
CTRL+Z terminates the process on Linux; we don't want that.

TTtie commented:
Switched to CTRL+G.  
Reasons:  
- CTRL + Z works on Windows, but not on Linux (process gets terminated)
- G for GAB :eyes:

Commit:
[Commands/About] About command 97% done

Commit:
Part 2 of the about command

Commit:
[Commands/alert] Alert command is done

Commit:
[*] Change line type from CRLF to LF

Commit:
[Web/Boot/Discord] /dashboard/management/shards
Shut GAwesomeBot down, restart GAwesomeBot!
Also, freeze, reset or restart any shard!

Commit:
[Bugs/Client] Add client to commands, and fix eval gist

Commit:
[Dependencies] Another update

Commit:
[Discord.js] Fix undefined

Commit:
[Base Commands] Add constants to the docs

Commit:
[Discord] Improvement, I guess

Commit:
[PMCommands] Fix bug gilbert found

Commit:
[WIP] Please ignroe

Commit:
[Web] Ooh, what is this?
Gilbert cleaning his working tree.

Commit:
Update maintainer perm defaults

With the evaluate permission, maintainers can "elevate" their perms to any level. Free execution of JavaScript enables users to have full control over the entirety of GAB, so it should be defaulted to The Host only. The shutdown commands are "shortcuts" of eval commands that sudo maintainers can have access to without undermining the permission system.

Commit:
[Web] Some whitespace improvements
And string updates!

Commit:
[Web] Woops
Turns out the blog/wiki composer/editor was always broken!

Commit:
[Web] Even more whitespacing updates

Commit:
[Github] Template updates

Commit:
[Cache] Fix bugs that Gilbert found
They're fixed! So thats good!

Commit:
[Commands] about command updates

Commit:
[Commands] Even more command updates

Commit:
[Discord] Oof, bug fixes!

Commit:
[Sharder] Mode option for the future

Commit:
[Web] Fix /userlist bug

Commit:
[Discord] They call me Gilbert, Dirty Gilbert

Commit:
[*] Clean Dirty Gilbert's.. Dirt

Commit:
[Structures] Fix my dumb

Commit:
[Discord] They call me Gilbert, Dirty Gilbert

Commit:
[Discord.js] Saw nothing

Commit:
[Trivia] Just me nitpicking.

Commit:
[Trivia] More nitpicks

This ones on me, I promise!

Commit:
[Eval] Fix some dumbs

Commit:
[Dependencies]

Commit:
[*] Move these

Commit:
[Cache/*] Fix cache issues

Commit:
[*] Gilbert shush

Commit:
Woopsie daisy you did NOT see this

Commit:
[DB] Shh

TTtie committed:
[Commands/CLI] leval

TTtie committed:
[Commands/CLI] leval

TTtie commented:
rebasing is a pain

TTtie committed:
Fix the conflicts

TTtie committed:
oops missed this

TTtie committed:
You didn't see this

TTtie committed:
[CLI] Don't write the arrow

TTtie committed:
[CLI/Commands] seval - the shard specific eval

TTtie committed:
[CLI/Commands] Use GAB's natural shutdown and restart

TTtie committed:
[CLI] Multiple changes; please look in the commit description for more info.
- Added a base CLI command so people can build their own.
- Put the db and traffic context so nobody has to dig it all up.

TTtie commented:
🤔 if I'll sure finish it til 2018

TTtie committed:
Update the files

TTtie committed:
move cli.setup(); to the shard finished event
this was made to not get the commands distorted (your input will be available after all gets ready.

TTtie committed:
addmaintainer cli command

TTtie committed:
delete the log because gab outputs it

TTtie committed:
Merge branch 'indev-4.0.2' of https://github.com/gilbertgobbels/gawesomebot into feature/indev-4.0.2/cli-commands

TTtie committed:
Don't add the trimmed data to multiline command input
May break some functionality in some cases.

TTtie committed:
However, trim the input for one-liners

TTtie committed:
eslint and Vlag Annoyer 9000 aka BigMessage

TTtie committed:
Various changes:
- switch to cli.sharder.broadcast()
- addmaintainer tells the shards to reload the config.json now
- added Vlag Annoyer 9000 aka BigMessage to the command list
- new IPC worker event: `updateConfigJSON`, used for ^

TTtie commented:
![obrazek](https://user-images.githubusercontent.com/17570452/34483615-fb9d6b3a-efbf-11e7-997a-99561766de5b.png)
This is how quality my commit descriptions are (No, we don't reload config.json on BigMessage from CLI)

TTtie committed:
whoops

TTtie committed:
removemaintainer command + use writeJSONAtomic

vladfrangu commented on Commands/CLI/addmaintainer.js:
```diff
@@ -0,0 +1,45 @@
+		return;
+	}
+	function isMaintainer (user, sudo) {
+		if (sudo === true) {
```
`if (sudo)`

vladfrangu commented on Commands/CLI/addmaintainer.js:
```diff
@@ -0,0 +1,45 @@
+	}
+	ids.forEach(user => {
+		if (isMaintainer(user, isSudo)) return;
+		if (isSudo === true && !isMaintainer(user, isSudo)) configJSON.sudoMaintainers.push(user);
```
`if (isSudo && !isMaintainer())`

vladfrangu commented on Commands/CLI/removemaintainer.js:
```diff
@@ -0,0 +1,45 @@
+		return;
+	}
+	function isMaintainer (user, sudo) {
+		if (sudo === true) {
```
Same as above

TTtie commented:
ok will do

TTtie commented:
Updated the title to use <!--shit~~-->git<!--~~-->moji.
<!-- If you saw that i called gitmoji shitmoji, it's literally because of my opinion which is based on "emoji in commit messages are ugly and bad". Thank you for understanding even though you shouldn't see this over the wire @ Discord. -->

TTtie commented:
hope no one burns for writing such a message

TTtie committed:
:twisted_rightwards_arrows: Update the repository

TTtie committed:
:ok_hand: Praise the Vlag

TTtie commented:
@KingDGrizzle @GilbertGobbels 🤔 
Should I make more of the stdin commands or uh.. consider this PR done

KingDGrizzle commented:
1. ConfigJSON spaces should be 2
2. I'd personally like some tests (and documentation?) for some commands
3. I'd like to see all commands in action 😄 

Otherwise, looks good enough for now. I'll let Gilbert say his opinion

TTtie committed:
:rotating_light: eslint

TTtie committed:
:ok_hand: Switch to 2 spaces

TTtie commented:
1. refactor is coming!
2. command help is enough?
3. 
```
$ git clone https://github.com/tttie/gawesomebot -b feature/indev-4.0.2/cli-commands
# fill the config.js and config.json like you normally would
$ node bot
# *gab gets ready*
# type your commands
```

TTtie committed:
:recycle: Refactor the stuff
The listing of a specific id is removed because of unlike the boot param, the logs don't get cluttered really much

TTtie committed:
:bug: Forgot to add the UpdateConfig require

TTtie commented:
@KingDGrizzle didn't ping you in the response, so have the ping here

vladfrangu commented on Modules/Utils/CLI/UpdateConfig.js:
```diff
@@ -0,0 +1,27 @@
+const { writeJSONAtomic: writeFile } = require("fs-nextra");
+
+/**Updates the config.json
```
should be on a new line.
Check other files that have JSDocs for examples

vladfrangu commented on Modules/Utils/CLI/UpdateConfig.js:
```diff
@@ -0,0 +1,27 @@
+ * @argument {Object} configJSON config.json object
+ * @returns {Promise}
+ */
+module.exports = function UpdateConfig(cli, configJSON) {
```
should be async (cli, configJSON), and return / throw instead of resolve/reject

vladfrangu commented on Modules/Utils/CLI/index.js:
```diff
@@ -0,0 +1,3 @@
+module.exports = {
```
This should be outside Utils

TTtie replied:
No, it actually shouldn't; these are *utilities* that can always be used by passsing a fake cli object (depends on the function)

vladfrangu commented on Modules/CLI.js:
```diff
@@ -0,0 +1,108 @@
+	/**
+	* Listens to the incoming events from stdin
+	* @param {Buffer} data The incoming data from the stdin
+	 * */
```
`* */`?
Should be `*/`

TTtie replied:
Do you mean `* blablaab\n*/`

vladfrangu replied:
yes

KingDGrizzle commented on Modules/CLI.js:
```diff
@@ -0,0 +1,108 @@
+	Errors: {
+		Error: GABError,
+	},
+} = require("../Internals/index");
```
Please add a new line :P

vladfrangu commented on Commands/CLI/removemaintainer.js:
```diff
@@ -0,0 +1,39 @@
+		if (i == (ids.length - 1)) return rs();
+	}));
+
+	updateCfg.then(() => UpdateConfig(cli, configJSON)).then(() => {
```
Could do with more async/await and trycatch

vladfrangu commented on Commands/CLI/addmaintainer.js:
```diff
@@ -0,0 +1,38 @@
+		winston.warn(`I need the IDs of the maintainers.`);
+		return;
+	}
+	function isMaintainer (user, sudo) {
```
Same as below? Above? Idk how GitHub will organize these

vladfrangu commented on Commands/CLI/bigmessage.js:
```diff
@@ -0,0 +1,5 @@
+module.exports = async ({ cli }, cmdData, args) => {
+	args = args.trim();
+	if (args.length === 0) return winston.error("No input for BigMessage");
```
`winston.warn` as error is used for critical erros

TTtie replied:
except that is winston.error, not warn

vladfrangu replied:
I said change it to `winston.warn` not it is winston.warn` rn

vladfrangu commented on Commands/CLI/seval.js:
```diff
@@ -0,0 +1,21 @@
+module.exports = async ({ cli }, cmdData, args) => {
+	if (!args) return winston.error("Missing arguments for seval");
```
Same as above, winston.warn

KingDGrizzle commented on Commands/CLI/seval.js:
```diff
@@ -0,0 +1,21 @@
+		shard = parseInt(shard);
+	}
+	let s;
+	if (code.length === 0) return winston.error("Missing code.");
```
As above

vladfrangu commented on Internals/Errors/Messages.js:
```diff
@@ -22,6 +22,19 @@ const Messages = {
 
+	// CLI errors
+	CLI_PARAM_INVALID: (paramName, receivedType, allowedTypes) => {
+		function mapTo (string) {
```
`const mapTo = string => `

TTtie replied:
show tttie wuv for not throwing es5 functions in trash

TTtie commented:
why must documentation be such a pain :/

vladfrangu commented on Modules/CLI.js:
```diff
@@ -0,0 +1,108 @@
+					this.commandArgs += strData;
+				}
+			}
+			delete this.currentlyRunningCommand;
```
`this.currentlyRunningCommand = null`

vladfrangu reviewed (changes requested):
Just a few things (cause Vlag loves nitpicks)

KingDGrizzle commented:
Also, please go through all eslint warnings for your files, and fix them 😄

TTtie commented:
my eslint found non3

KingDGrizzle commented:
It's called Travis Builds ;)

TTtie commented:
duh ~~spaces are the future!~~

TTtie committed:
:rotating_light: eslint

TTtie committed:
:ok_hand: Fixing jsdocs

TTtie committed:
:ok_hand: Un-nitpick the nitpicks

TTtie committed:
:ok_hand: Move the asterisks

KingDGrizzle commented:
Minus some ES5 / ES6? Combo, LGTM
@GilbertGobbels for final review

TTtie commented:
gilbort pls 😴

Gilbert142 commented on Configurations/commands.js:
```diff
@@ -957,6 +957,50 @@ const commands = {
+			isMultiline: true,
+		},
+
+		addmaintainer: {
```
Merge addmaintainer and removemaintainer into one, name it "promote", and simply check if the ID's specified already exist as maintainers or not, then act accordingly.

TTtie replied:
Hmm, looks even harder than expected; the fact I have to check so it doesn't remove a sudo maintainer that I'm promoting from normal maintainer list.

TTtie replied:
Because ~~I'm lazy enough to find a solution~~ can't really make a solution on my side it would be easier to merge them, name the command `maintainer` and make them choose if the said maintainer is going to be added or removed.

Gilbert142 commented on Discord.js:
```diff
@@ -1441,6 +1441,15 @@ bot.IPC.on("restart", async (msg, callback) => {
 	}
 });
 
+bot.IPC.on("updateConfigJSON", async (_, callback) => {
```
`reloadConfig` as the event name would make more sense, also, you can still call the first parameter `msg`, it would be less confusing for contributors.

vladfrangu replied:
The first param is not needed, hence why it's called `_`. It's ignored

Gilbert142 replied:
I fully understand the usage, I just disagree with it. It isn't consistent compared to the other IPC callbacks and naming it semanticly just makes it look nicer too.

Gilbert142 commented on bot.js:
```diff
@@ -110,6 +110,9 @@ database.initialize(configJS.databaseURL).catch(err => {
 		const traffic = new Traffic(sharder.IPC, false);
 		traffic.init();
 
+		winston.verbose("Creating a STDIN CLI instance.");
```
Can be winston.silly

Gilbert142 commented on Modules/Utils/CLI/UpdateConfig.js:
```diff
@@ -0,0 +1,26 @@
+ * @argument {Object} configJSON config.json object
+ * @returns {Promise<Boolean|Error>}
+ */
+module.exports = async function UpdateConfig (cli, configJSON) {
```
When not using `this`, please try to use fat arrow functions at all times `() => {}`

Gilbert142 reviewed (dismissed):
I still have to test this but from a quick glance, these are the issues I could find so far. There's some refactoring to do here and there, but we're getting there!

TTtie committed:
:twisted_rightwards_arrows: Merge the GAB

TTtie committed:
:ok_hand: Update the event name and use an arrow function for updating configuration

TTtie committed:
:ok_hand: Use winston.silly

TTtie committed:
:ok_hand: use msg instead of _ parameter

TTtie committed:
update gab

TTtie committed:
:recycle: :fire: Dependency-less config.json reload + merge maintainer commands into one

TTtie commented:
@GilbertGobbels hey gilbort, what about now?

TTtie commented:
heck

TTtie commented:
currently working on fixing my mistakes, be right back

TTtie committed:
:rotating_light: :fire: eslint + removal of some debugging code, attempt no. 2

TTtie committed:
Merge branch 'indev-4.0.2' of https://github.com/gilbertgobbels/gawesomebot into feature/indev-4.0.2/cli-commands

vladfrangu commented on Commands/CLI/promote.js:
```diff
@@ -0,0 +1,52 @@
+		ids.shift();
+	}
+
+	if (ids.length === 0) {
```
`if (!ids.length)`

vladfrangu commented on Commands/CLI/promote.js:
```diff
@@ -0,0 +1,52 @@
+	await updateCfg;
+	try {
+		await UpdateConfig(cli, configJSON);
+		winston.info(`Promoted/demoted users to/from ${isSudo ? "Sudo" : ""} Maintainer`);
```
Can't you make the string based on `action`?

TTtie replied:
file size is the key vlag.

vladfrangu commented on Commands/CLI/restart.js:
```diff
@@ -0,0 +1,4 @@
+const { spawn } = require("child_process");
```
Unused import

vladfrangu commented on Commands/CLI/seval.js:
```diff
@@ -0,0 +1,21 @@
+		shard = parseInt(shard);
+	}
+	let s;
+	if (code.length === 0) return winston.error("Missing code.");
```
Move the check above `let s` and also, `if (!code.length)`
Also, winston.warn

vladfrangu commented on Commands/CLI/seval.js:
```diff
@@ -0,0 +1,21 @@
+	}
+	let s;
+	if (code.length === 0) return winston.error("Missing code.");
+	if (!cli.sharder.shards.get(shard)) {
```
`.has` instead of `.get`

TTtie replied:
There's no point to change it in my opinion. Both can return a falsy value, so 🤔

vladfrangu replied:
Yes, but `has` checks just if the value is present, while `get` also returns it.

vladfrangu reviewed (changes requested):
More code style stuff that I'd like changed.

TTtie committed:
:ok_hand: Try satisfying the god

TTtie committed:
Merge branch 'indev-4.0.2' of https://github.com/gilbertgobbels/gawesomebot into feature/indev-4.0.2/cli-commands

vladfrangu commented on Commands/CLI/promote.js:
```diff
@@ -0,0 +1,56 @@
+module.exports = async ({ cli }, cmdData, args) => {
+	args = args.trim();
+	const [action, ...ids] = args.split(" ");
+	if (action !== "add" || action !== "remove") {
```
By the by sweetheart, toLowerCase :D

TTtie committed:
:ok_hand: vlag isn't the god

TTtie committed:
:fire: let's save some space

KingDGrizzle commented:
@GilbertGobbels please take one final look, otherwise, LGTM

TTtie commented:
@GilbertGobbels, a free reminder to review this because Vlag told you to exactly 20 days ago.

TTtie committed:
:twisted_rightwards_arrows: Update GAB

TTtie committed:
:bug: Add a listener only if stdin is present

TTtie commented:
@GilbertGobbels
1. Another reminder to review this
2. This isnt WIP anymore, plz remove the in progress label

Gilbert142 reviewed (changes requested):
Please move Modules/CLI.js to Internals/CLI.js

TTtie committed:
Merge branch 'indev-4.0.2' of https://github.com/gilbertgobbels/gawesomebot into feature/indev-4.0.2/cli-commands

TTtie committed:
Move CLI into internals

TTtie committed:
:rotating_light: Remove a tab

Gilbert142 commented on Internals/CLI.js:
```diff
@@ -0,0 +1,110 @@
+		} else {
+			if (this.currentlyRunningCommand.data.isMultiline) {
+				if (trimmedData.endsWith(`\u0007`)) {
+					trimmedData = trimmedData.replace(/\u0007/, "");
```
This line seems to be making an eslint error occur, eventhough this shouldn't have to be regex (You don't use the global flag, so this can be a simple string rather than a Regex)

Either turn this into a string, or fix the regex (and, if desired, add the global flag)

TTtie replied:
oh wait i found out one mistake by myself. This regex. Misses $.

EDIT 18:12 CET: Or maybe not, you're right. I think I will rather set up an ignore because in the console it can be seen as `^G`

TTtie replied:
@GilbertGobbels If disabling the eslint error in 3080e46 shouldn't be done, let me know.

TTtie committed:
:rotating_light: Disable the no-control-regex rule

wasdennnoch commented on Commands/CLI/help.js:
```diff
@@ -0,0 +1,13 @@
+		return `${cmd.isMultiline ? "*" : ""}${c}${cmd.usage ? ` ${cmd.usage}` : ""}${cmd.description ? ` - ${cmd.description}` : ""}`;
+	});
+	winston.info(" ======================== COMMAND LIST ========================");
+	str.forEach(s => winston.info(s));
```
str.forEach(winston.info); :P

TTtie replied:
No. [Array.forEach()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) also calls the function passed with the item index and the array itself, which isn't a good idea.

wasdennnoch replied:
Oh right. Disregard.

TTtie committed:
:ok_hand: Tell the user the message is sending

TTtie committed:
:rotating_light: semicolon :/

TTtie commented:
Just wondering if this is still going to be a thing. It's been half a year already :eyes:

KingDGrizzle commented:
Rebase + lint

TTtie commented:
Lint..  check.
Merge.. not yet.

TTtie committed:
Merge branch 'indev-4.0.2' of https://github.com/gilbertgobbels/gawesomebot into feature/indev-4.0.2/cli-commands

TTtie committed:
:rotating_light: Fix the pesky eslint rules
I had to donate my HDD space for another two plugins too!

TTtie commented:
Okay. All should be OK now.

TTtie committed:
Merge branch 'indev-4.0.2' of https://github.com/gilbertgobbels/gawesomebot into feature/indev-4.0.2/cli-commands

TTtie commented:
omg its half a year and I didnt add anything useful to this

i guess i will have to merge the og source into this again

TTtie committed:
Update the GAB after half a year, please review if everything works

TTtie committed:
remove the tabs frm L119

TTtie commented:
AFAIK, GAB next will be containerized - is this still on the roadmap or shall I close this off?
